include/swift/SILOptimizer/Utils/UpdatingInstructionIterator.h

```diff
@@ -0,0 +1,243 @@
+#include <utility>
+
+namespace swift {
+
+class UpdatingInstructionIteratorRegistry;
+
+namespace detail {
+
+struct UpdatingInstructionIteratorState {
+  using Self = UpdatingInstructionIteratorState;
+
+  SILBasicBlock::iterator base;
+  bool isAdvanced = false;
+  unsigned refCount = 1;
+
+  UpdatingInstructionIteratorState(SILBasicBlock::iterator base) : base(base) {}
+
+  void advance() {
+    if (isAdvanced)
+      isAdvanced = false;
+    else
+      ++base;
+  }
+
+  void notifyDelete(SILInstruction *inst) {
+    if (base == inst->getIterator()) {
+      isAdvanced = true;
+      ++base;
+    }
+  }
+
+  void notifyNew(SILInstruction *inst) {
+    if (isAdvanced && std::prev(base) == inst->getIterator())
+      --base;
+  }
+
+  /// This is undefined on the block's end iterator.
+  SILInstruction *getInstruction() const {
+    return isAdvanced ? nullptr : &*base;
+  }
+
+  // Exclude reference counting from equality.
+  bool operator==(const Self &rhs) const {
+    return this->base == rhs.base && this->isAdvanced == rhs.isAdvanced;
+  }
+  bool operator!=(const Self &rhs) const { return !(*this == rhs); }
+};
+
+} // namespace detail
+
+/// Safely iterate over instructions while deleting instructions and inserting
+/// new instructions.
+class UpdatingInstructionIterator {
+  using Self = UpdatingInstructionIterator;
+  using State = detail::UpdatingInstructionIteratorState;
+
+  UpdatingInstructionIteratorRegistry *registry;
+  unsigned id = 0;
+
+  State &state();
+  const State &state() const;
+
+  void retain() const;
+  void release() const;
+
+public:
+  using value_type = SILInstruction *;
+  using difference_type = std::ptrdiff_t;
+  using pointer = SILInstruction **;
+  using reference = SILInstruction *&;
+  using iterator_category = std::bidirectional_iterator_tag;
+
+  UpdatingInstructionIterator(UpdatingInstructionIteratorRegistry &registry,
+                              unsigned id)
+      : registry(&registry), id(id) {}
+
+  UpdatingInstructionIterator(const Self &rhs)
+      : registry(rhs.registry), id(rhs.id) {
+    retain();
+  }
+
+  Self &operator=(const Self &rhs) {
+    rhs.retain();
+    release();
+
+    this->registry = rhs.registry;
+    this->id = rhs.id;
+    return *this;
+  }
+
+  ~UpdatingInstructionIterator() { release(); }
+
+  /// This returns nullptr for deleted instructions.
+  SILInstruction *operator*() const { return state().getInstruction(); }
+
+  SILInstruction *operator->() const { return operator*(); }
+
+  Self &operator++() {
+    state().advance();
+    return *this;
+  }
+
+  bool operator==(const Self &rhs) const { return state() == rhs.state(); }
+  bool operator!=(const Self &rhs) const { return !(*this == rhs); }
+};
+
+/// Track iterators that need updating when intructions are added or
+/// deleted. Iterators can be tracked across multiple levels of the call
+/// stack. This registry object must outlive any iterators that it vends.
+class UpdatingInstructionIteratorRegistry {
```

Comment:
Are you expecting this to be used outside InstructionDeleter?
If not I suggest to merge this with InstructionDeleter (to keep things simple).

Reply:
The implementations of InstructionDeleter and UpdatingInstructionIteratorRegistry are unrelated and each complex in their own way. I assure you it's much simpler in a separate class. I do imagine moving it out of InstructionDeleter at some point.

Reply:
ok